keras_cv/layers/transformer_encoder.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class TransformerEncoder(layers.Layer):
+    """
+    Transformer encoder block implementation as a Keras Layer.
+
+    Args:
+        project_dim: the dimensionality of the projection of the encoder
+        intermediate_dim: default 768, the intermediate dimensionality of the transformer encoder
+        mlp_dim: default 3072, the intermediate dimensionality of the MLP head before projecting to `project_dim`
+        num_heads: the number of heads for the `MultiHeadAttention` layer
+        mlp_dropout: default 0.1, the dropout rate to apply between the layers of the MLP head of the encoder
+        attention_dropout: default 0.1, the dropout rate to apply in the MultiHeadAttention layer
+        activation: default tf.nn.gelu(), the activation function to apply in the MLP head
+        layer_norm_epsilon: default 1e-06, the epsilon for `LayerNormalization` layers
+
+    Basic usage:
+
+    ```
+    project_dim = 1024
+    intermediate_dim = 768
+    mlp_dim = 3072
+    num_heads = 4
+
+    patches = keras_cv.layers.Patching(patch_size)(cat_img)
+    encoded_patches = keras_cv.layers.PatchEmbedding(project_dim=project_dim)(patches)
+    trans_encoded = keras_cv.layers.TransformerEncoder(project_dim=project_dim,
+                                                       intermediate_dim=intermediate_dim,
+                                                       mlp_dim = mlp_dim,
+                                                       num_heads=num_heads)(encoded_patches)
+
+    print(trans_encoded.shape) # (1, 197, 1024)
+    ```
+    """
+
+    def __init__(
+        self,
+        project_dim,
+        num_heads,
+        intermediate_dim=768,
+        mlp_dim=3072,
```

Comment:
just curious -- is this the default value being used in ViT? 3072 seems a really high dimension to me.

Reply:
good point; 

we probably don't want a default for either of these given that this is meant to be a generic layer.

Reply:
Yup, a pretty high dimensionality. They set it as the dims for their "base" model for the output of the MLP head in the param tables.

Let me re-read the paper to check if I got it right or not. Even if I did, should we maybe reduce it?

Reply:
MLP dims for ViT-Base:

<img width="405" alt="image" src="https://user-images.githubusercontent.com/60978046/200823275-424eac58-b6b2-4cab-aa7f-94cfcc53fb75.png">

HuggingFace's ViT config also uses 3072 as the default value:

https://huggingface.co/docs/transformers/model_doc/vit#transformers.ViTConfig

Reply:
It _feels_ wrong, but it looks like it's right haha

Reply:
Though, since it's a generic transformer encoder - we might as well have it be 512 or 1024 for other use cases and then supply 3072 for ViTs specifically?

Reply:
If someone wants to build a custom ViTs or ViT derivatives with these, it makes sense to keep the defaults for ViTs though so that they don't have to update the args?
Not sure what's the right choice...

DeiT and BeiT have 3072 too. It might be better to have the defaults stay as 768 and 3072 because they seem to be a common default. I totally get why we'd want to decrease them too though, so whatever you feel like is the right choice works

Reply:
We can either go with having defaults for all of these (12 heads is also a common default) or go with none of them

Reply:
DPT (Dense Prediction Transformer) for CV also have 768 hidden and 3072 mlp dims

Reply:
TL:DR;

ViT, DeiT, BeiT, DPT, ViTMAE, ViTMSE use 768 (transformer hidden dims), 3072 (MLP dims) and 12 heads as the defaults/lower bounds for models.
LeViT and SwinT married convolutions back into the mix and use lower dimensionality for the MLP heads and transformer encoders because of the lower dims coming in (I think).

We can go either with what they've "historically" been (quotes because they're all from the past 2 years) or anticipate more SwinT-type models that use convs and lower dims and remove the defaults fully. Thoughts?

Reply:
> Thoughts?

Theoretically it could be ok but I will find hard to communicate to the user base what they are really using when they adopt/retain/fine-tune Keras-cv models.

I mean that generally users expect to have the same arch when we are using "well known names" and we need to handle customization optionally or handling it with a specific care.

Reply:
We can do that when building models. The `MODELS_CONFIG` would have dims set to 768/3072 for ViTs for example, totally regardless of the default transformer args.
The question is whether we want to set the encoder defaults to something or nothing at all?

Reply:
Yes I was talking more about the 2nd part:
> or anticipate more SwinT-type models that use convs and lower dims and remove the defaults fully.

Reply:
Maybe we go with no defaults for the generic transformer layer, and enforce the common dimensionality on a per-model basis, since we'll have those set _anyway_

Reply:
i think having no defaults is a better way to communicate to users what is going on.  I tend to prefer explicitness when any ambiguity is involved.

Reply:
Makes sense! Removed the defaults :)